src/libstd/sys/wasm/mod.rs

```diff
@@ -102,3 +96,205 @@ pub unsafe fn abort_internal() -> ! {
+            let mut call_record = ArgsSysCall {
+                len: buffer.len(),
+                ptr: buffer.as_mut_ptr(),
+                result: 0
+            };
+            unsafe { syscall(SysCallIndex::Args, &mut call_record); }
+            Ok(call_record.result)
+        })
+            .unwrap()
+            .split(|b| *b == 0)
+            .map(|s| FromInner::from_inner(Buf { inner: s.to_owned() }))
+            .collect()
+    }
+}
+
+#[repr(C)]
+pub struct GetEnvSysCall {
+    key_ptr: *const u8,
+    key_len: usize,
+    value_ptr: *mut u8,
+    value_len: usize,
+    result: usize
+}
+
+impl GetEnvSysCall {
+    pub fn perform(key: &OsStr) -> Option<OsString> {
+        let key_buf = &AsInner::as_inner(key).inner;
+        receive_buffer(64, |buffer| {
+            let mut call_record = GetEnvSysCall {
+                key_len: key_buf.len(),
+                key_ptr: key_buf.as_ptr(),
+                value_len: buffer.len(),
+                value_ptr: buffer.as_mut_ptr(),
+                result: !0usize
+            };
+            unsafe { syscall(SysCallIndex::GetEnv, &mut call_record); }
+
+            if call_record.result == !0usize {
+                Err(())
+            } else {
+                Ok(call_record.result)
+            }
+        }).ok().map(|s| {
+            FromInner::from_inner(Buf { inner: s })
+        })
+    }
+}
+
+#[repr(C)]
+pub struct SetEnvSysCall {
+    key_ptr: *const u8,
+    key_len: usize,
+    value_ptr: *const u8,
+    value_len: usize
+}
+
+impl SetEnvSysCall {
+    pub fn perform(key: &OsStr, value: Option<&OsStr>) {
+        let key_buf = &AsInner::as_inner(key).inner;
+        let value_buf = value.map(|v| &AsInner::as_inner(v).inner);
+        let mut call_record = SetEnvSysCall {
+            key_len: key_buf.len(),
+            key_ptr: key_buf.as_ptr(),
+            value_len: value_buf.map(|v| v.len()).unwrap_or(!0usize),
+            value_ptr: value_buf.map(|v| v.as_ptr()).unwrap_or(ptr::null())
+        };
+        unsafe { syscall(SysCallIndex::SetEnv, &mut call_record); }
+    }
+}
+
+pub enum TimeClock {
+    Monotonic = 0,
+    System = 1,
+}
+
+#[repr(C)]
+pub struct TimeSysCall {
+    clock: usize,
+    secs_hi: usize,
+    secs_lo: usize,
+    nanos: usize
+}
+
+impl TimeSysCall {
+    pub fn perform(clock: TimeClock) -> Duration {
+        let mut call_record = TimeSysCall {
+            clock: clock as usize,
+            secs_hi: 0,
+            secs_lo: 0,
+            nanos: 0
+        };
+        unsafe { syscall(SysCallIndex::Time, &mut call_record); }
+        if call_record.secs_hi != 0 || call_record.secs_lo != 0 || call_record.nanos != 0 {
+            Duration::new(
+                ((call_record.secs_hi as u64) << 32) | (call_record.secs_lo as u64),
+                call_record.nanos as u32
+            )
+        } else {
+            panic!("Time system call is not implemented by WebAssembly host");
+        }
```

Comment:
Assuming we would go with the syscall interface, shouldn't we make it possible for the JS side to signal "not implemented" in a more generic way for all of the syscalls instead of having it hardcoded for only this one?

For example, we could have the `rust_wasm_syscall` return an integer to signify whenever it handled a given syscall: `0` for unhandled, `1` for handled.

Reply:
That assumes that you want not-implemented syscalls to panic: this differs from the current behaviour.

At the moment all system calls are designed such that a no-op implementation will give the most sensible fallback: for time the fallback should be a panic, because there's no way to gracefully degrade. For reading/writing it's possible to degrade without panicking.

I think the logic about whether to panic or degrade has to sit on the rust side, because we want to put as little burden as possible on the user. Given that, I don't think having a specific return value buys us anything?

The only motivating example would be if the user wants to always panic on unimplemented syscalls regardless of whether it's possible to degrade gracefully. In that case the user can raise an exception on the javascript side. At the moment wasm uses panic=abort, so panicking is also a javascript exception. Alternatively, the user can export the `rust_wasm_syscall` from their own rust code, and implement a "true" panic.

Reply:
> That assumes that you want not-implemented syscalls to panic [...] I think the logic about whether to panic or degrade has to sit on the rust side

Actually what I said doesn't assume we want the unimplemented syscalls to panic. I think you're right that the logic whenever to panic should be on Rust's side, in which case we **don't** want to hardcode in the interface itself which syscalls will and which will not panic.

A generic interface to signal which syscalls are implemented and which are not gives Rust the option to decide what panics and what not, and lets it change its mind in the future with no change to the interfaces.

Reply:
Oh I see, you just mean for consistency? Does an undefined return value from an import get coerced to false in wasm?

Reply:
It gets converted to zero; WASM doesn't have booleans yet in function prototypes.

See [here](https://webassembly.github.io/spec/js-api/index.html#towebassemblyvalue):

> The algorithm ToWebAssemblyValue(v, type) coerce a JavaScript value to a WebAssembly value performs the following steps:
>
> Assert: type is not 𝗂𝟨𝟦.
>    * If type is 𝗂𝟥𝟤,
>       1. Let i32 be ? ToInt32(v).
>       2. Return 𝗂𝟥𝟤.𝖼𝗈𝗇𝗌𝗍 i32.
>    * If type is 𝖿𝟥𝟤,
>        1. Let f32 be ? ToNumber(v) rounded to the nearest representable value using IEEE 754-2008 round to nearest, ties to even mode.
>        2. Return 𝖿𝟥𝟤.𝖼𝗈𝗇𝗌𝗍 f32.
>    * If type is 𝖿𝟨𝟦,
>        1. Let f64 be ? ToNumber(v).
>        2. Return 𝖿𝟨𝟦.𝖼𝗈𝗇𝗌𝗍 f64.

So basically to convert a JS type to WASM `i32` applies [ToInt32](https://tc39.github.io/ecma262/#sec-toint32):

> The abstract operation ToInt32 converts argument to one of 2^32 integer values in the range -2^31 through 2^31 - 1, inclusive. This abstract operation functions as follows:
> 
>    1. Let number be ? ToNumber(argument).
>    2. If number is NaN, +0, -0, +∞, or -∞, return +0.
>    3. Let int be the mathematical value that is the same sign as number and whose magnitude is floor(abs(number)).
>    4. Let int32bit be int modulo 232.
>    5. If int32bit ≥ 231, return int32bit - 232; otherwise return int32bit. 

And [ToNumber](https://tc39.github.io/ecma262/#sec-tonumber) converts an `undefined` to `NaN`, which per `ToInt32` gets converted to a zero.

Reply:
Thanks, that's a really useful link!